File: app/http/endpoints/api/analytics/staffdetail.go
package api_analytics

import (
	"context"
	"net/http"
	"strconv"
	"time"

	"github.com/TicketsBot-cloud/dashboard/app"
	dbclient "github.com/TicketsBot-cloud/dashboard/database"
	"github.com/TicketsBot-cloud/dashboard/log"
	"github.com/TicketsBot-cloud/dashboard/rpc/cache"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"
)

type staffDetailResponse struct {
	UserId               uint64                  `json:"user_id,string"`
	Username             string                  `json:"username"`
	Avatar               string                  `json:"avatar"`
	FirstResponseTime    tripleWindowSeconds     `json:"first_response_time"`
	AverageRating        *float32                `json:"average_rating"`
	RatingCount          int                     `json:"rating_count"`
	FeedbackDistribution [5]int                  `json:"feedback_distribution"`
	TicketsClaimed       staffDetailTripleWindow `json:"tickets_claimed"`
	TicketsAnswered      staffDetailTripleWindow `json:"tickets_answered"`
	GuildTotalTickets    staffDetailTripleWindow `json:"guild_total_tickets"`
	OpenClaimedCount     int                     `json:"open_claimed_count"`
}

type staffDetailTripleWindow struct {
	AllTime int `json:"all_time"`
	Monthly int `json:"monthly"`
	Weekly  int `json:"weekly"`
}

func GetAnalyticsStaffDetailHandler(ctx *gin.Context) {
	guildId := ctx.Keys["guildid"].(uint64)

	userIdStr := ctx.Param("userid")
	userId, err := strconv.ParseUint(userIdStr, 10, 64)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	var resp staffDetailResponse
	resp.UserId = userId

	var frtAllTime, frtMonthly, frtWeekly *time.Duration

	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	group, groupCtx := errgroup.WithContext(timeoutCtx)

	// First response time - all time
	group.Go(func() (err error) {
		frtAllTime, err = dbclient.Client.FirstResponseTime.GetAverageAllTimeUser(groupCtx, guildId, userId)
		return
	})

	// First response time - monthly (28 days)
	group.Go(func() (err error) {
		frtMonthly, err = dbclient.Client.FirstResponseTime.GetAverageUser(groupCtx, guildId, userId, time.Hour*24*28)
		return
	})

	// First response time - weekly (7 days)
	group.Go(func() (err error) {
		frtWeekly, err = dbclient.Client.FirstResponseTime.GetAverageUser(groupCtx, guildId, userId, time.Hour*24*7)
		return
	})

	// Average rating
	group.Go(func() error {
		avg, err := dbclient.Client.ServiceRatings.GetAverageClaimedBy(groupCtx, guildId, userId)
		if err != nil {
			return err
		}
		if avg > 0 {
			resp.AverageRating = &avg
		}
		return nil
	})

	// Rating count
	group.Go(func() error {
		count, err := dbclient.Client.ServiceRatings.GetCountClaimedBy(groupCtx, guildId, userId)
		if err != nil {
			return err
		}
		resp.RatingCount = count
		return nil
	})

	// Feedback distribution
	group.Go(func() (err error) {
		resp.FeedbackDistribution, err = dbclient.Client.ServiceRatings.GetDistributionClaimedBy(groupCtx, guildId, userId)
		return
	})

	// Tickets claimed - all time
	group.Go(func() error {
		count, err := dbclient.Client.TicketClaims.GetClaimedCount(groupCtx, guildId, userId)
		if err != nil {
			return err
		}
		resp.TicketsClaimed.AllTime = count
		return nil
	})

	// Tickets claimed - monthly
	group.Go(func() error {
		count, err := dbclient.Client.TicketClaims.GetClaimedSinceCount(groupCtx, guildId, userId, time.Hour*24*28)
		if err != nil {
			return err
		}
		resp.TicketsClaimed.Monthly = count
		return nil
	})

	// Tickets claimed - weekly
	group.Go(func() error {
		count, err := dbclient.Client.TicketClaims.GetClaimedSinceCount(groupCtx, guildId, userId, time.Hour*24*7)
		if err != nil {
			return err
		}
		resp.TicketsClaimed.Weekly = count
		return nil
	})

	// Tickets answered - all time
	group.Go(func() error {
		count, err := dbclient.Client.Participants.GetParticipatedCount(groupCtx, guildId, userId)
		if err != nil {
			return err
		}
		resp.TicketsAnswered.AllTime = count
		return nil
	})

	// Tickets answered - monthly
	group.Go(func() error {
		count, err := dbclient.Client.Participants.GetParticipatedCountInterval(groupCtx, guildId, userId, time.Hour*24*28)
		if err != nil {
			return err
		}
		resp.TicketsAnswered.Monthly = count
		return nil
	})

	// Tickets answered - weekly
	group.Go(func() error {
		count, err := dbclient.Client.Participants.GetParticipatedCountInterval(groupCtx, guildId, userId, time.Hour*24*7)
		if err != nil {
			return err
		}
		resp.TicketsAnswered.Weekly = count
		return nil
	})

	// Guild total tickets - all time
	group.Go(func() error {
		count, err := dbclient.Client.Tickets.GetTotalTicketCount(groupCtx, guildId)
		if err != nil {
			return err
		}
		resp.GuildTotalTickets.AllTime = count
		return nil
	})

	// Guild total tickets - monthly
	group.Go(func() error {
		count, err := dbclient.Client.Tickets.GetTotalTicketCountInterval(groupCtx, guildId, time.Hour*24*28)
		if err != nil {
			return err
		}
		resp.GuildTotalTickets.Monthly = count
		return nil
	})

	// Guild total tickets - weekly
	group.Go(func() error {
		count, err := dbclient.Client.Tickets.GetTotalTicketCountInterval(groupCtx, guildId, time.Hour*24*7)
		if err != nil {
			return err
		}
		resp.GuildTotalTickets.Weekly = count
		return nil
	})

	// Open claimed count
	group.Go(func() error {
		count, err := dbclient.Client.TicketClaims.GetOpenClaimedCount(groupCtx, guildId, userId)
		if err != nil {
			return err
		}
		resp.OpenClaimedCount = count
		return nil
	})

	if err := group.Wait(); err != nil {
		log.Logger.Error("Failed to retrieve staff detail analytics", zap.Uint64("guild_id", guildId), zap.Uint64("user_id", userId), zap.Error(err))
		_ = ctx.AbortWithError(http.StatusInternalServerError, app.NewError(err, "Failed to retrieve staff analytics. Please try again later."))
		return
	}

	resp.FirstResponseTime = tripleWindowSeconds{
		AllTime: durationToSeconds(frtAllTime),
		Monthly: durationToSeconds(frtMonthly),
		Weekly:  durationToSeconds(frtWeekly),
	}

	// Resolve username from cache
	users, err := cache.Instance.GetUsers(timeoutCtx, []uint64{userId})
	if err != nil {
		log.Logger.Warn("Failed to resolve user data for staff detail", zap.Error(err))
	} else if u, ok := users[userId]; ok {
		resp.Username = u.Username
		resp.Avatar = u.AvatarUrl(256)
	}

	ctx.JSON(200, resp)
}

File: frontend/src/lib/analytics-export/staff-detail.ts
import { appendCsvSection } from "@/lib/csv";
import { formatDuration } from "@/lib/analytics-format";
import type { StaffDetailAnalytics } from "@/types";
import type { ExportMeta, ExportSection } from "./types";

export const STAFF_DETAIL_EXPORT_SECTIONS: ExportSection[] = [
  { id: "summary", label: "Summary stats" },
  { id: "first_response_time", label: "First response time" },
  { id: "tickets_claimed", label: "Tickets claimed" },
  { id: "tickets_answered", label: "Tickets answered vs guild total" },
  { id: "feedback_distribution", label: "Feedback distribution" },
];

export interface StaffDetailExportData {
  detail: StaffDetailAnalytics;
  guildId: string;
}

export function buildStaffDetailMeta(data: StaffDetailExportData): ExportMeta {
  return {
    exported_at: new Date().toISOString(),
    guild_id: data.guildId,
    user_id: data.detail.user_id,
    username: data.detail.username,
  };
}

function buildSectionPayload(id: string, data: StaffDetailExportData): unknown {
  const { detail } = data;

  switch (id) {
    case "summary":
      return {
        average_rating: detail.average_rating,
        rating_count: detail.rating_count,
        avg_first_response_seconds: detail.first_response_time?.weekly,
        avg_first_response_formatted: formatDuration(detail.first_response_time?.weekly ?? null),
        tickets_claimed_all_time: detail.tickets_claimed.all_time,
        tickets_claimed_monthly: detail.tickets_claimed.monthly,
        open_claimed_count: detail.open_claimed_count,
      };
    case "first_response_time":
      return {
        weekly_seconds: detail.first_response_time?.weekly,
        weekly_formatted: formatDuration(detail.first_response_time?.weekly ?? null),
        monthly_seconds: detail.first_response_time?.monthly,
        monthly_formatted: formatDuration(detail.first_response_time?.monthly ?? null),
        all_time_seconds: detail.first_response_time?.all_time,
        all_time_formatted: formatDuration(detail.first_response_time?.all_time ?? null),
      };
    case "tickets_claimed":
      return {
        weekly: detail.tickets_claimed.weekly,
        monthly: detail.tickets_claimed.monthly,
        all_time: detail.tickets_claimed.all_time,
      };
    case "tickets_answered":
      return {
        weekly: {
          answered: detail.tickets_answered.weekly,
          guild_total: detail.guild_total_tickets.weekly,
        },
        monthly: {
          answered: detail.tickets_answered.monthly,
          guild_total: detail.guild_total_tickets.monthly,
        },
        all_time: {
          answered: detail.tickets_answered.all_time,
          guild_total: detail.guild_total_tickets.all_time,
        },
      };
    case "feedback_distribution":
      return (detail.feedback_distribution ?? [0, 0, 0, 0, 0]).map((count, idx) => ({
        stars: idx + 1,
        count,
      }));
    default:
      return null;
  }
}

export function buildStaffDetailPayload(
  selectedIds: string[],
  data: StaffDetailExportData,
): Record<string, unknown> {
  const meta = buildStaffDetailMeta(data);
  const sections: Record<string, unknown> = {};

  for (const id of selectedIds) {
    sections[id] = buildSectionPayload(id, data);
  }

  return { ...meta, sections };
}

export function buildStaffDetailCsv(selectedIds: string[], data: StaffDetailExportData): string {
  const { detail } = data;
  const parts: string[] = [];

  for (const id of selectedIds) {
    switch (id) {
      case "summary":
        appendCsvSection(
          parts,
          "Summary Stats",
          ["Metric", "Value", "Notes"],
          [
            [
              "Avg Rating",
              detail.average_rating != null ? `${detail.average_rating.toFixed(1)}/5` : "No data",
              `${detail.rating_count} responses`,
            ],
            [
              "Avg First Response",
              formatDuration(detail.first_response_time?.weekly ?? null),
              detail.first_response_time?.weekly,
            ],
            [
              "Tickets Claimed",
              detail.tickets_claimed.all_time,
              `${detail.tickets_claimed.monthly} this month`,
            ],
            ["Open Tickets", detail.open_claimed_count, ""],
          ],
        );
        break;
      case "first_response_time":
        appendCsvSection(
          parts,
          "First Response Time",
          ["Period", "Duration", "Seconds"],
          [
            [
              "Weekly",
              formatDuration(detail.first_response_time?.weekly ?? null),
              detail.first_response_time?.weekly,
            ],
            [
              "Monthly",
              formatDuration(detail.first_response_time?.monthly ?? null),
              detail.first_response_time?.monthly,
            ],
            [
              "All Time",
              formatDuration(detail.first_response_time?.all_time ?? null),
              detail.first_response_time?.all_time,
            ],
          ],
        );
        break;
      case "tickets_claimed":
        appendCsvSection(
          parts,
          "Tickets Claimed",
          ["Period", "Count"],
          [
            ["Weekly", detail.tickets_claimed.weekly],
            ["Monthly", detail.tickets_claimed.monthly],
            ["All Time", detail.tickets_claimed.all_time],
          ],
        );
        break;
      case "tickets_answered":
        appendCsvSection(
          parts,
          "Tickets Answered",
          ["Period", "Answered", "Guild Total"],
          [
            ["Weekly", detail.tickets_answered.weekly, detail.guild_total_tickets.weekly],
            ["Monthly", detail.tickets_answered.monthly, detail.guild_total_tickets.monthly],
            ["All Time", detail.tickets_answered.all_time, detail.guild_total_tickets.all_time],
          ],
        );
        break;
      case "feedback_distribution":
        appendCsvSection(
          parts,
          "Feedback Distribution",
          ["Rating", "Count"],
          (detail.feedback_distribution ?? [0, 0, 0, 0, 0]).map((count, idx) => [
            `${idx + 1} star`,
            count,
          ]),
        );
        break;
    }
  }

  return parts.join("\r\n");
}
